core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2110,4 +2125,192 @@ public void testCurrentTranslogIDisCommitted() throws IOException {
+            Engine.Index index = new Engine.Index(newUid("1"), doc);
+            engine.index(index);
+
+            // Fill the listener slots
+            List<DummyRefreshListener> nonForcedListeners = new ArrayList<>(INDEX_SETTINGS.getMaxRefreshListeners());
+            for (int i = 0; i < defaultSettings.getMaxRefreshListeners(); i++) {
+                DummyRefreshListener listener = new DummyRefreshListener(index.getTranslogLocation());
+                nonForcedListeners.add(listener);
+                engine.addRefreshListener(listener);
+            }
+
+            // We shouldn't have called any of them
+            for (DummyRefreshListener listener : nonForcedListeners) {
+                assertNull("Called listener too early!", listener.forcedRefresh.get());
+            }
+
+            // Add one more listener which should cause a refresh. In this thread, no less.
+            DummyRefreshListener forcingListener = new DummyRefreshListener(index.getTranslogLocation());
+            engine.addRefreshListener(forcingListener);
+            assertTrue("Forced listener wasn't forced?", forcingListener.forcedRefresh.get());
+
+            // That forces all the listeners through. On the listener thread pool so give them some time with assertBusy.
+            for (DummyRefreshListener listener : nonForcedListeners) {
+                assertBusy(
+                        () -> assertEquals("Expected listener called with unforced refresh!", Boolean.FALSE, listener.forcedRefresh.get()));
+            }
+        }
+    }
+
+    public void testAddRefreshListenerAfterRefresh() throws Exception {
+        try (Store store = createStore();
+                Engine engine = createEngine(defaultSettings, store, createTempDir(), NoMergePolicy.INSTANCE)) {
+            ParsedDocument doc = testParsedDocument("1", "1", "test", null, -1, -1, testDocumentWithTextField(), B_1, null);
+            Engine.Index index = new Engine.Index(newUid("1"), doc);
+            engine.index(index);
+            engine.refresh("I said so");
+            if (randomBoolean()) {
+                engine.index(new Engine.Index(newUid("1"), doc));
+                if (randomBoolean()) {
+                    engine.refresh("I said so");
+                }
+            }
+
+            DummyRefreshListener listener = new DummyRefreshListener(index.getTranslogLocation());
+            engine.addRefreshListener(listener);
+            assertFalse(listener.forcedRefresh.get());
+        }
+    }
+
+    /**
+     * Uses a whole bunch of threads to index, wait for refresh, and non-realtime get documents to validate that they are visible after
+     * waiting regardless of what crazy sequence of events causes the refresh listener to fire.
+     */
+    public void testAddRefreshListenerLotsOfThreads() throws Exception {
+        int threadCount = between(5, max(50, defaultSettings.getMaxRefreshListeners() * 2));
+        long runTime = TimeUnit.SECONDS.toNanos(5);
```

Comment:
first of all you get time dependent and reproducibilty will suck bigtime since if it depends on the number of operations or some buffers etc it will never reproduce on other machines. also your test will always sleep which sucks since they might sleep for nothing?